src/Adapter/Product/AdminProductWrapper.php

```diff
@@ -493,19 +493,40 @@ public function processProductCustomization($product, $data)
 
         $shopList = Shop::getContextListShopID();
 
+        /** Update the customization fields to be deleted in the next step */
+        $updateQuery = 'UPDATE `'._DB_PREFIX_.'customization_field` cf
+            INNER JOIN ' . _DB_PREFIX_ . 'customized_data cd ON cd.`index` <> cf.`id_customization_field`
+            SET cf.`is_deleted` = 0 
+            WHERE 
+            cf.`is_deleted` = 1 ';
+        Db::getInstance()->execute($updateQuery);
+
+        /** update deleted customization */
+        $updateQuery = 'UPDATE `'._DB_PREFIX_.'customization_field` cf
+            INNER JOIN ' . _DB_PREFIX_ . 'customized_data cd ON cd.`index` = cf.`id_customization_field` 
+            SET cf.`is_deleted` = 1 
+            WHERE 
+            cf.`is_deleted` = 0 
+            AND cf.`id_product` = ' . (int)$product->id;
+        if (!empty($customization_ids)) {
+            $updateQuery .= ' AND cf.`id_customization_field` NOT IN (' . implode(',', array_map('intval', $customization_ids)) . ')';
+        }
+        Db::getInstance()->execute($updateQuery);
+
         //remove customization field langs for current context shops
         foreach ($product->getCustomizationFieldIds() as $customizationFiled) {
             //if the customization_field is still in use, only delete the current context shops langs,
             //else delete all the langs
             Db::getInstance()->execute('DELETE FROM '._DB_PREFIX_.'customization_field_lang WHERE
 			`id_customization_field` = '.(int)$customizationFiled['id_customization_field'].
+            ' AND 0 <> (SELECT count(*) FROM `'._DB_PREFIX_.'customization_field` WHERE `is_deleted` = 0 AND `id_customization_field` = ' . (int)$customizationFiled['id_customization_field'] . ')' .
             (in_array((int)$customizationFiled['id_customization_field'], $customization_ids) ? ' AND `id_shop` IN ('.implode(',', $shopList).')' : ''));
         }
 
         //remove unused customization for the product
         if (!empty($customization_ids)) {
             Db::getInstance()->execute('DELETE FROM '._DB_PREFIX_.'customization_field WHERE
-            `id_product` = '.(int)$product->id.' AND `id_customization_field` NOT IN ('.implode(",", $customization_ids).')');
+            `is_deleted` = 0 AND `id_product` = '.(int)$product->id.' AND `id_customization_field` NOT IN ('.implode(",", $customization_ids).')');
```

Comment:
Here too, please move "is_deleted ..." after "id_product ..." to ensure that MySQL will use the index

Reply:
Are we absolutely, 100% certain that `$customization_ids` contains only integers? Otherwise this is open to SQL injection.

Reply:
Even if that case, we suggest in the modules to call `array_map()` on the array in order to apply `intval()`.

Reply:
Yes it contains only integers https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/Product/AdminProductWrapper.php#L493